protocols/relay/src/transport.rs

```diff
@@ -0,0 +1,432 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::behaviour::{BehaviourToTransportMsg, OutgoingRelayReqError};
+use crate::protocol;
+use crate::RequestId;
+use futures::channel::mpsc;
+use futures::channel::oneshot;
+use futures::future::{BoxFuture, Future, FutureExt};
+use futures::sink::SinkExt;
+use futures::stream::{Stream, StreamExt};
+use libp2p_core::either::{EitherError, EitherFuture, EitherOutput};
+use libp2p_core::multiaddr::{Multiaddr, Protocol};
+use libp2p_core::transport::{ListenerEvent, TransportError};
+use libp2p_core::{PeerId, Transport};
+use pin_project::pin_project;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+
+pub enum TransportToBehaviourMsg {
+    DialReq {
+        request_id: RequestId,
+        relay_addr: Multiaddr,
+        relay_peer_id: PeerId,
+        dst_addr: Multiaddr,
+        dst_peer_id: PeerId,
+        send_back: oneshot::Sender<
+            Result<protocol::Connection, OutgoingRelayReqError>,
+        >,
+    },
+    ListenReq {
+        address: Multiaddr,
+        peer_id: PeerId,
+    },
+}
+
+/// A [`Transport`] wrapping another [`Transport`] enabling relay capabilities.
+///
+/// Allows the local node to:
+///
+/// 1. Establish relayed connections by dialing `/p2p-circuit` addresses.
+/// 2. Accept incoming relayed connections.
+/// 3. Connecting to relay nodes in order for them to listen for incoming
+///    connections for the local node.
+#[derive(Clone)]
+pub struct RelayTransportWrapper<T: Clone> {
+    to_behaviour: mpsc::Sender<TransportToBehaviourMsg>,
+    from_behaviour: Arc<Mutex<mpsc::Receiver<BehaviourToTransportMsg>>>,
```

Comment:
Does it not matter which transport instance emits the `ListenerEvent::Upgrade` received through this channel? If I understand correctly, then through sharing the channel receiver in this manner, if I create a transport and clone it to "listen on" two different relay nodes, I can get events for "inbound connections" from one reported by the other? At least I would expect to get only those from a listener that actually came from the relay that I "listened on" with this particular transport instance. Do I misunderstand the implementation? It even looks like I can get events for inbound relayed connections on a listener that did not even "listen on" a relay address.

Reply:
> Do I misunderstand the implementation? It even looks like I can get events for inbound relayed connections on a listener that did not even "listen on" a relay address.

Your understanding above is correct.

One needs to support two kinds of incoming relayed connections:
1. Incoming relayed connections relayed via a known relay node, for which the local node previously explicitly called `Swarm::listen_on(&mut swarm, RELAY_ADDR + CIRCUIT_RELAY_POSTFIX).unwrap();` i.e. explicitly asked the relay node to act as a relay for the local node.
2. Incoming relayed connections via some (unspecified) relay node, via which the local node is not explicitly listening for incoming connections.

While I would consider use-case (1) the most important one, e.g. to overcome NATs, I still think the use-case (2) is relevant. Say e.g. that node _A_ and _B_ want to connect. Both are publicly routable, but don't have a encryption protocol in common. In case there is a relay node R which has an encryption protocol in common with both _A_ and _B_, the relay node can act as a mediator between the two nodes _A_ and _B_.

Thus far, I did not come up with a clean interface which would support both (1) and (2). As you correctly noticed, for now I simply (/naively) pass an incoming relayed connection to the first listener polling the channel to the `Relay` behaviour, ignoring whether the listener is actually listening on a relay connection. I agree that this is not ideal.

Properly supporting use-case (1) and always passing an incoming relayed connection to the listener listening via that specific relay node should be easily doable via additional channels routing an incoming relayed connection to the correct listener.

I am not yet sure how to support use-case (2). First thing that comes to my mind is to ask users, that want to receive incoming relayed connections via any relay node, to register a _catch-all_ listener, e.g. via `Swarm::listen_on(&mut swarm, "/p2p-circuit".try_into().unwrap()).unwrap();` which would then yield all incoming relayed connections that are relayed via a relay node via which the local node is not explicitly listening.

Does the above make sense? If so, can you think of a better interface to support both use-cases?

Reply:
> First thing that comes to my mind is to ask users, that want to receive incoming relayed connections via any relay node, to register a catch-all listener, e.g. via Swarm::listen_on(&mut swarm, "/p2p-circuit".try_into().unwrap()).unwrap(); which would then yield all incoming relayed connections that are relayed via a relay node via which the local node is not explicitly listening.

That sounds good to me!